src/Body.ts

```diff
@@ -1,44 +1,590 @@
-import WidgetBase from '@dojo/widget-core/WidgetBase';
-import { HasColumns, HasItems } from './interfaces';
+import WidgetBase, { onPropertiesChanged } from '@dojo/widget-core/WidgetBase';
+import { HasBufferRows, HasColumns, HasItems, HasScrollTo, HasScrollToEvent, HasSize, HasSlice, HasSliceEvent, ItemProperties } from './interfaces';
 import Row from './Row';
 
-import * as bodyClasses from './styles/body.m.css';
+import * as css from './styles/body.m.css';
 
 export const BodyBase = ThemeableMixin(RegistryMixin(WidgetBase));
 
-export interface BodyProperties extends ThemeableProperties, HasColumns, HasItems, RegistryMixinProperties { }
+export interface BodyProperties extends ThemeableProperties, HasBufferRows, HasColumns, HasItems, HasScrollTo, HasScrollToEvent, HasSize, HasSlice, HasSliceEvent, RegistryMixinProperties {}
 
-@theme(bodyClasses)
+interface RenderedDetails {
+	add: boolean;
+	remove: boolean;
+	element?: HTMLElement;
+	index: number;
+	height?: number;
+}
+
+@theme(css)
 class Body extends BodyBase<BodyProperties> {
-	render() {
+	private _firstVisibleKey: string;
+	private _itemElementMap = new Map<string, RenderedDetails>();
+	private _marginTop?: RenderedDetails;
+	private _scrollerElement: HTMLElement;
+	private _scrollTop = 0;
+
+	/**
+	 * Creates a DNode for the passed item and either
+	 * creates or updates an associated entry in
+	 * the item element map.
+	 */
+	protected createNodeFromItem(item: ItemProperties, index: number) {
 		const {
-			columns,
-			items,
-			registry,
-			theme
-		} = this.properties;
+			_itemElementMap: itemElementMap,
+			properties: {
+				columns,
+				theme,
+				registry
+			}
+		} = this;
+
+		const key = item.id;
+		let details = itemElementMap.get(key);
+		if (!details) {
+			details = {
+				add: true,
+				remove: false,
+				index: index
+			};
+			itemElementMap.set(key, details);
+		}
+		else {
+			details.index = index;
+		}
 
 		return v('div', {
-				classes: this.classes(bodyClasses.scroller)
-			},
-			[
-				v('div', {
-					classes: this.classes(bodyClasses.content)
+			key
+		}, [
+			w<Row>('row', {
+				columns,
+				item,
+				key,
+				registry,
+				theme
+			})
+		]);
+	}
+
+	/**
+	 * Uses the height of the scroll area and the estimated row height
+	 * to estimate the number of rows that will fill it
+	 */
+	protected estimatedRowCount(): number {
+		const {
+			_scrollerElement: scrollerElement
+		} = this;
+
+		const height = scrollerElement.offsetHeight;
+		if (height) {
+			return Math.round(height / this.estimatedRowHeight());
+		}
+		return 30;
+	}
+
+	/**
+	 * Based on what is currently rendered, find the average height
+	 * of each row and, if nothing is rendered, return 20.
+	 */
+	protected estimatedRowHeight(): number {
+		const {
+			_itemElementMap: itemElementMap
+		} = this;
+
+		let rowHeight = 0;
+		let rowCount = 0;
+		for (let details of from(itemElementMap.values())) {
```

Comment:
Does itemElementMap contain collapsed rows in a tree scenario, and would they have an element that could mess up the average height if that is the case? Not sure if we're even worried about tree stuff for this PR.